File: frontend/src/types/album.ts
/**
 * Albums — a user-named COLLECTION of projects (`backend/app/schemas/album.py`).
 *
 * ★ MANY-TO-MANY, and that is the whole point of the entity. A project belongs to as
 *   many albums as its owner likes — "Spring 2026", "Client: Northgate", "Coastal" —
 *   the way a track belongs to several playlists. An album is therefore NOT an owner
 *   and NOT a folder: deleting one detaches its projects and destroys nothing. Every
 *   membership call (`POST`/`DELETE /albums/{id}/projects/{project_id}`) is
 *   IDEMPOTENT for the same reason — a multi-toggle UI must be free to re-assert a
 *   state it already holds.
 *
 * ★ THE CASING LAW (§8.1 / L9): snake_case, mirroring the wire, no mapping layer.
 */

import type { IsoDateTime, Uuid } from './common';

/**
 * ★ The membership reference EMBEDDED IN `ProjectSummary.albums` — the reason a
 *   project card can render its album chips without a second request per row. It is
 *   deliberately thin: an id, a name and the accent colour are everything a chip
 *   needs, and inlining `project_count` here would make every project list pay for a
 *   rollup nothing on that screen reads.
 */
export interface AlbumRef {
  id: Uuid;
  name: string;
  color?: string | null;
}

/**
 * `GET /albums` row. ★ `project_count` is a server-side rollup — counting client-side
 * would need every album's membership list, i.e. one request per card.
 */
export interface AlbumSummary {
  id: Uuid;
  /** ★ CASE-INSENSITIVELY UNIQUE among live albums → `409 ALBUM_NAME_CONFLICT`. */
  name: string;
  description: string | null;
  /** A CSS hex colour (`#rrggbb`) used as the card's accent, or `null` for the default. */
  color: string | null;
  project_count: number;
  created_at: IsoDateTime;
  updated_at: IsoDateTime;
}

/** `GET /albums/{id}`, and the body of every album write. */
export interface Album extends AlbumSummary {
  /** Albums are soft-deleted, like projects; a live one is always `null`. */
  deleted_at?: IsoDateTime | null;
}

export interface AlbumCreate {
  name: string;
  description?: string | null;
  color?: string | null;
}

/**
 * PATCH — ★ UNSET-SENTINEL SEMANTICS (§6.1), and here it is load-bearing twice over:
 *   - an omitted key leaves the field untouched, `null` CLEARS it;
 *   - **an entirely empty body is a `400 EMPTY_PATCH`**, so callers must send only
 *     the fields that actually changed and must not submit an unchanged form.
 *
 * `JSON.stringify` drops `undefined` keys, so the `T | null | undefined` typing does
 * this for free — do not "normalise" the body.
 */
export interface AlbumUpdate {
  name?: string | undefined;
  description?: string | null | undefined;
  color?: string | null | undefined;
}

/** `GET /albums` query. `sort` accepts `name`, `created_at`, `project_count`, `-` prefixed. */
export interface AlbumFilters {
  q?: string;
  sort?: string;
  limit?: number;
  offset?: number;
}

File: backend/app/schemas/album.py
"""Albums — a user-defined, **many-to-many** collection of projects.

Mirrors ``project.py``'s shape exactly: a ``*Create``, a ``*Update`` on
:class:`~app.schemas.common.PatchModel` with UNSET semantics, a fat ``*Read`` and a
thin ``*Summary`` list projection.

★ **The membership endpoints carry no body at all** —
``POST/DELETE /albums/{album_id}/projects/{project_id}`` say everything in the path and
answer ``204``. There is therefore no ``AlbumProjectCreate`` here: a body whose only
possible content is the id already in the URL is a second place for the two to disagree.

★ :class:`~app.schemas.common.AlbumRef` — the ``{id, name}`` pair a project row carries —
lives in ``common`` rather than here, for the reason §6.2 gives when it moved
``RevisionSummary`` there: ``project.py`` needs it and this module needs it, and a
``project → album`` import would be a fourth domain-to-domain edge in a package whose
stated rule is one-way ``common``/``enums``/``errors`` ← everything else.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import Field

from .common import ApiModel, ListParams, PatchModel

__all__ = [
    "ALBUM_SORT_FIELDS",
    "AlbumCreate",
    "AlbumListParams",
    "AlbumRead",
    "AlbumSummary",
    "AlbumUpdate",
]

#: ★ ``project_count`` is on the whitelist and is **not a column** — the repository
#: resolves it onto a correlated count over ``album_projects``, exactly as
#: ``PROJECT_SORT_FIELDS`` does for ``image_count``. Every sort is stabilised by the
#: repository with a trailing ``id ASC``.
ALBUM_SORT_FIELDS = frozenset({"name", "created_at", "updated_at", "project_count"})

#: 1–200, non-blank after strip. ``str_strip_whitespace`` on ``ApiModel`` runs BEFORE
#: ``min_length``, so ``"   "`` becomes ``""`` and fails the bound with no extra
#: validator — the same mechanism ``ProjectName`` relies on, and it mirrors
#: ``ck_albums_name_nonblank``.
AlbumName = Annotated[str, Field(min_length=1, max_length=200)]

#: A hex triple/sextet or a short design token (``"#3B82F6"``, ``"amber"``). Deliberately
#: **not** a strict hex pattern: the field is a UI tint, the frontend owns the palette,
#: and rejecting ``"amber"`` here would make the server the arbiter of a colour system it
#: cannot see. The bound mirrors ``ck_albums_color_short`` so a violation is a clean 422
#: rather than a ``23514`` from Postgres.
AlbumColor = Annotated[str, Field(min_length=1, max_length=32)]


class AlbumCreate(ApiModel):
    """``POST /albums``.

    ``{"name": "2026 season"}`` alone is a complete album — description and colour are
    both genuinely optional, and neither is invented server-side.
    """

    name: AlbumName
    description: str | None = Field(default=None, max_length=4000)
    color: AlbumColor | None = Field(
        default=None,
        description=(
            "A short hex or design token for UI tinting, e.g. '#3B82F6' or 'amber'. "
            "★ null means NO COLOUR CHOSEN — the server never picks one."
        ),
    )


class AlbumUpdate(PatchModel):
    """``PATCH /albums/{album_id}`` — rename / re-describe / re-tint. ★ UNSET semantics.

    ``{"color": null}`` **clears** the colour; ``{}`` is a ``400 EMPTY_PATCH`` raised by
    the router off :meth:`PatchModel.is_empty`. ``name`` is ``T | None`` only because
    every PATCH field must be omittable — clearing it is refused by the service, since
    an album with no name is not a thing the UI can render.
    """

    name: AlbumName | None = None
    description: str | None = Field(default=None, max_length=4000)
    color: AlbumColor | None = None


class AlbumRead(ApiModel):
    """``AlbumRead`` — the create, get and patch responses."""

    id: UUID
    name: str
    description: str | None
    color: str | None
    project_count: int = Field(
        ge=0,
        description=(
            "How many projects are in this album. ★ One batched aggregate, never a "
            "loaded relationship — ``Album.projects`` is ``lazy='raise'`` precisely so "
            "an N+1 cannot creep in through it."
        ),
    )
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class AlbumSummary(ApiModel):
    """List projection — ``GET /albums``.

    Carries the same fields as :class:`AlbumRead` minus ``deleted_at``: an album has no
    expensive projection to omit (no AOI, no five-subquery rollup), so the two differ
    only in the soft-delete marker, which a list of live albums has no use for.
    """

    id: UUID
    name: str
    description: str | None
    color: str | None
    project_count: int = Field(ge=0)
    created_at: datetime
    updated_at: datetime


class AlbumListParams(ListParams):
    """``GET /albums``."""

    q: str | None = Field(default=None, description="Free-text over name + description.")
    include_deleted: bool = False
